src/libraries/Microsoft.Extensions.Primitives/tests/StringSegmentTest.cs

```diff
@@ -37,9 +37,75 @@ public void StringSegment_AsSpan()
         {
             var segment = new StringSegment("Hello");
 
-            var span = segment.AsSpan();
+            ReadOnlySpan<char> span = segment.AsSpan();
 
-            Assert.Equal(5, span.Length);
+            Assert.Equal("Hello", span.ToString());
+        }
+
+        [Fact]
+        public void StringSegment_AsSpanStart_Valid()
+        {
+            var segment = new StringSegment("Hello, World!", 1, 4);
+
+            ReadOnlySpan<char> result = segment.AsSpan(start: 1);
+
+            Assert.Equal("llo", result.ToString());
+        }
+
+        [Fact]
+        public void StringSegment_AsSpan_Valid()
+        {
+            var segment = new StringSegment("Hello, World!", 1, 4);
+
+            ReadOnlySpan<char> result = segment.AsSpan(start: 1, length: 2);
+
+            Assert.Equal("ll", result.ToString());
+        }
+
+        [Fact]
+        public void StringSegment_AsSpan_Empty()
+        {
+            var segment = new StringSegment("Hello, World!");
+
+            ReadOnlySpan<char> result = segment.AsSpan(start: 1, length: 0);
+
+            Assert.True(result.IsEmpty);
+        }
+
+        [Fact]
+        public void StringSegment_AsSpan_InvalidOffset()
+        {
+            var segment = new StringSegment("Hello, World!", 1, 3);
+
+            ArgumentOutOfRangeException exception = Assert.Throws<ArgumentOutOfRangeException>(() => segment.AsSpan(-1, 1));
+            Assert.Equal("start", exception.ParamName);
+        }
+
+        [Fact]
+        public void StringSegment_AsSpan_InvalidLength()
+        {
+            var segment = new StringSegment("Hello, World!", 1, 3);
+
+            ArgumentOutOfRangeException exception = Assert.Throws<ArgumentOutOfRangeException>(() => segment.AsSpan(0, -1));
+            Assert.Equal("length", exception.ParamName);
+        }
+
+        [Fact]
+        public void StringSegment_AsSpan_InvalidOffsetAndLength()
+        {
+            var segment = new StringSegment("Hello, World!", 1, 3);
+
+            ArgumentException exception = Assert.Throws<ArgumentException>(() => segment.AsSpan(2, 3));
+            Assert.Contains("bounds", exception.Message);
+        }
+
+        [Fact]
+        public void StringSegment_AsSpan_OffsetAndLengthOverflows()
```

Comment:
How about a test or two for `!HasValue`?

Reply:
Added with 4e4a7663b09192c7786a28cce4ebb8238f46a84b